### DISCUSSION TITLE
How to merge partial configuration `ocelot.*.json` files on the fly?

### DISCUSSION THREAD

**Discussion by SheruGaur** [ThreeMammals/Ocelot#2042] [Q&A] [answered]
We are trying to merge multiple `ocelot.json` files without restarting the application with below code. However we are successfully able to merge on starting the application, but still application restart is required to merge multiple `ocelot.*.json` files changes. Even we updated latest ocelot version **[23.2.2](https://github.com/ThreeMammals/Ocelot/releases/tag/23.2.2)**.

Please guide if we are missing any reference or any mistake in below mentioned code.

```csharp
webBuilder.ConfigureAppConfiguration((context, config) =>
{
    var folder = Path.Combine(AppDomain.CurrentDomain.BaseDirectory);
    var path = Path.Combine(folder, "My\\folder");
    config.AddOcelot(path, context.HostingEnvironment, MergeOcelotJson.ToFile, optional: false, reloadOnChange: true); // with folder, environment and merging type
})
```
Our `ocelot.json` file structure is like:
  1. ocelot.common.json
  2. ocelot.mfcommon.json

**Comment by raman-m** [comment, marked as answer]
Hi @SheruGaur 
#### The Q 1
> still application restart is required to merge multiple `ocelot.*.json` files changes. 

Correct! If you have partial config files, they will be merged only during Ocelot app start. And the same configuration will be available during Docker container life time.

#### The Q 2
> On the fly `ocelot.json` configuration merging

This feature is not implemented for partial config files, `ocelot.*.json`!
Ocelot is able to track `ocelot.json` on the fly if you use [Administration](https://ocelot.readthedocs.io/en/latest/features/administration.html) feature or copy the `ocelot.json` file to app folder manually.
**Merging partial files on the fly requires a lot of development❕** 

--------------
I have a questions for you.
- Why do you need to update partial config files so often ?
- What's your user scenario?
- What's your deployment environment: self-contained, IIS, Docker, K8s, cloud?

If your environment is Docker and/or K8s then why not to manage Ocelot app version and its configuration `ocelot.json` by Docker image versioning, and create/kill Docker containers?
This is more easy DevOps approach than juggling multiple partial files for long living Ocelot app instance.

**Reply by SheruGaur** [reply]
@raman-m  Thank you for your quick response.
 
> **Why do you need to update partial config files so often?**

Answer: The reason to go for the multiple file (partial) config is that we are managing multiple product configuration and we found it easier to manage the configuration as per the names of the file each file contains the information for the configuration for the particular products.

> **What's your user scenario?**

Answer: For exposing multiple product endpoints through single entry point.

> **What's your deployment environment: self-contained, IIS, Docker, K8s, cloud?**

Answer: Docker

However, we understood what you are trying to explain. But we are still in some confusion regarding the partial file configuration on fly merging as mentioned in the documentation of the ocelot for your references 👉
- [Merging Configuration Files](https://ocelot.readthedocs.io/en/latest/features/configuration.html#merging-configuration-files)

Exactly specifying the points of confusion in section: [Reload JSON Config On Change](https://ocelot.readthedocs.io/en/latest/features/configuration.html#reload-json-config-on-change)
>Ocelot supports reloading the JSON configuration file on change. For instance, the following will recreate Ocelot internal configuration when the `ocelot.json` file is updated manually:

#### Could you please guide on the same?
Thanks, and Waiting for your response.

**Reply by raman-m** [reply]
Shery,
The lifespan of your Docker container will not be a topic of discussion, as it falls outside the Ocelot team's purview.
I am firmly convinced that versioning Docker images is preferable to manipulating configuration JSON files.

As for the confusion...
> But we are still in some confusion regarding the partial file configuration on fly merging as mentioned in the documentation of the ocelot for your references

Could you point out the specific text that states Ocelot can merge configurations "on the fly"?
On the other hand, we have yet to clarify to developers that the [Merging Configuration Files](https://ocelot.readthedocs.io/en/latest/features/configuration.html#merging-configuration-files) feature is operational only during the application's startup. Thus...
> Ocelot will then save the merged configuration to a file called [ocelot.json](https://github.com/ThreeMammals/Ocelot/blob/main/test/Ocelot.ManualTest/ocelot.json) and this will be used as the source of truth while Ocelot is running.

Firstly, an experienced ASP.NET developer would infer from these words that the configuration remains static post-merging and startup.
However, I acknowledge that this might perplex some junior ASP.NET developers.
For instance, are you aware that the `ConfigureAppConfiguration` method is invoked solely during the startup of an ASP.NET web application?
Once the application has started, you cannot call the `ConfigureAppConfiguration` method, nor can you employ the merging feature within `AddOcelot`.
To me, this is quite straightforward, and I am puzzled by your confusion.

Secondly, it's important to recognize that the features [Merging Configuration Files](https://ocelot.readthedocs.io/en/latest/features/configuration.html#merging-configuration-files) and [Reload JSON Config On Change](https://ocelot.readthedocs.io/en/latest/features/configuration.html#reload-json-config-on-change) are distinct. The former is a native Ocelot feature for the application's startup phase, while the latter is an ASP.NET framework feature. They should not be conflated into a single feature.

Given that "on the fly" merging is not currently implemented in Ocelot and is only available during the startup phase, here are my recommendations:
- Employ the [Merging Configuration Files](https://ocelot.readthedocs.io/en/latest/features/configuration.html#merging-configuration-files) feature when you need to merge multiple configuration partial files during the Ocelot application's startup phase only. Utilize Docker image versioning to distribute changes in partial configuration files.
- Opt for the [Reload JSON Config On Change](https://ocelot.readthedocs.io/en/latest/features/configuration.html#reload-json-config-on-change) feature if **partial files are not used**, and there is a single `ocelot.json` with a few environmental variants. This approach is better suited for long-running Ocelot instances.

### To conclude
#### If you require the merging of partial client configuration files "on the fly," consider contributing to our project.

**Reply by raman-m** [reply]
@SheruGaur
I've prepared PR #2048, which includes developer's notes on utilizing the [Merging Configuration Files](https://ocelot.readthedocs.io/en/latest/features/configuration.html#merging-configuration-files) feature.
Hopefully, this will clear up any confusion for you and the other developers. 😉

**Reply by raman-m** [reply]
New updated Configuration doc is here: [configuration.rst | Merging Configuration Files](https://github.com/ThreeMammals/Ocelot/blob/develop/docs/features/configuration.rst#merging-configuration-files), see **Note 2**!
